Field Slips

mo-nathan opened the pull request:
This provides the key functionality needed for Field Slips.  See #1947 for the issue and #1896 for a more detailed discussion.

This PR delivers the functionality where there is a route that accepts a field slip code.  The /qr/:code route is provided for simplicity and brevity.  To test you probably want to update some projects with field slip prefixes.  For example, if you're working with a recent database dump you could add the prefix '6F' to the [First Fantastic Fungal Friends Foray Festival!](http://localhost:3000/projects/336) or 'NFAL' to [North Falmouth Fungi](http://localhost:3000/projects/334?q=1pkXE).  You can then exercise the new functionality with URLs like:

localhost:3000/qr/6F-0001
localhost:3000/qr/NFAL-0001

The #1947 has a checklist of functionality that I added as part of the PR to get to the point where it feels like a cohesive whole.

If you want to try it starting from a QR code, you can create QR codes here: https://www.the-qrcode-generator.com/

mo-nathan committed:
Ran:
rails generate scaffold FieldSlip observation_id:integer project_id:integer identifier:string

mo-nathan committed:
Switch from identifier to code and add routes

mo-nathan committed:
Add field_slips table

mo-nathan committed:
Rubocop

mo-nathan committed:
Enable field slip show to take a code and to take the user to the
new field slip page with the appropriate code if the code doesn't exist
yet.

mo-nathan committed:
Require login for FieldSlip actions other than show

mo-nathan committed:
Initial show field slips improvements

nimmolo commented on app/views/field_slips/_form.html.erb:
```diff
@@ -0,0 +1,32 @@
+      </ul>
+    </div>
+  <% end %>
+
```
I think in this situation we've been using `<div class="alert alert-danger">` (instead of `<div style="color: red">`)

nimmolo commented on app/views/field_slips/_form.html.erb:
```diff
@@ -0,0 +1,32 @@
+    <%= form.label :code, style: "display: block" %>
+    <%= form.text_field :code %>
+  </div>
+
```
The helper `text_field_with_label` should produce the Bootstrap markup here that we use elsewhere. 

We may need a new `number_field_with_label` helper to do that for number fields.

mo-nathan replied:
Adopted.  Thanks!

nimmolo replied:
Glad it helped. I think you can drop the wrapping `<div>`s too. The Bootstrap form element classes will handle it.

nimmolo commented on app/views/field_slips/show.html.erb:
```diff
@@ -0,0 +1,10 @@
+  <%= link_to "Back to field slips", field_slips_path %>
+
+  <%= button_to "Destroy this field slip", @field_slip, method: :delete %>
+</div>
```
Idk if you'd want the index field slips to also show these "edit/destroy" links per field slip, but it might be possible to pass a "context" local, and put those icon-links, plus a "back to index" link `if context == :show`... Something like that.

Can provide syntax for the icon links if you want to do it that way.

mo-nathan replied:
I don't think this is needed so I'm not going to sweat it.

mo-nathan commented:
@nimmolo Thanks for the comments.  Happy to follow your suggestions.  Most of the code is actually from `rails generate scaffold`.  I was just trying to get something working as a starting point.  Did you see my comment in #general in Slack?  I'm trying to figure out how to make the matrix_box in app/views/field_slips/_field_slip.html.erb look reasonable instead of ultra skinny.  I'm guessing there's some reasonably simple way to do that.  I played around with it a bit but I'm hopeless with front end stuff.

mo-nathan committed:
Add field_slip_prefix to Projects

mo-nathan committed:
Merge branch 'main' into njw-field-slips

mo-nathan committed:
Allow for empty string input for field_slip_prefix

mo-nathan committed:
Field slip ERB and language tag cleanup

mo-nathan committed:
Add field_slip_prefix migration

nimmolo commented on app/views/field_slips/_field_slip.html.erb:
```diff
@@ -0,0 +1,16 @@
+
+  <%= render(partial: "shared/matrix_box",
+             locals: { object: field_slip.observation }) %>
+</div>
```
I think this is all you need here. (The row width might need adjusting.)
This is from collection_numbers#show
```ruby
<ul class="row list-unstyled mt-3">
  <%= render(partial: "shared/matrix_box", locals: { columns: "col-xs-12" },
             collection: @collection_number.observations, as: :object) %>
</ul>
```

mo-nathan replied:
Found a solution based on the herbarium record that is working well.

mo-nathan committed:
Test the tests passing

mo-nathan committed:
Avoid showing a success box if there are no notices

mo-nathan committed:
Coverage and edge cases

mo-nathan committed:
Hookup a FieldSlip to a Project based on the field slip prefix

mo-nathan committed:
Merge branch 'main' into njw-field-slips

mo-nathan committed:
/qr/ route for field_slips#show

mo-nathan committed:
Add . and - to qr regex

mo-nathan committed:
Merge branch 'main' into njw-field-slips

mo-nathan committed:
Add Field Slip to Observation#show page

mo-nathan committed:
Enable edit/create to lead to create workflow or to use last viewed observation. Show/index cleanup.

mo-nathan committed:
Add thumbail of last image

mo-nathan committed:
Merge branch 'main' into njw-field-slips

mo-nathan committed:
String and test cleanup

mo-nathan committed:
Add user_id to FieldSlips

mo-nathan committed:
Add User to FieldSlips

mo-nathan committed:
Edit field slip can keep current image. Test coverage.

mo-nathan committed:
Use the FieldSlip code for default_accession_number

mo-nathan committed:
Add project drop down to field slip form

mo-nathan committed:
Add the field slip project to create checkboxes plus some more coverage

mo-nathan committed:
Automatically add users to open projects when they create an associated field slip

mo-nathan committed:
Typo

mo-nathan committed:
Rubocop on app/views/field_slips/show.json.jbuilder

mo-nathan committed:
Switch to with_label helpers and fixed a bug

mo-nathan committed:
Rubocop

mo-nathan committed:
Rubocop

JoeCohen commented on test/system/field_slips_test.rb:
```diff
@@ -0,0 +1,47 @@
+
+class FieldSlipsTest < ApplicationSystemTestCase
+  setup do
+    @field_slip = field_slips(:one)
```
```suggestion
    @field_slip = field_slips(:field_slip_one)
```
Fixes 
```ruby
ERROR FieldSlipsTest#test_visiting_the_index (0.75s)
Minitest::UnexpectedError:         StandardError: No fixture named 'one' found for fixture set 'field_slips'
            test/system/field_slips_test.rb:7:in `block in <class:FieldSlipsTest>'
```

mo-nathan replied:
Weird.  The tests pass fine for me locally without this change which I totally don't understand.  Thanks for fixing it.

mo-nathan replied:
Oh, I thought you pushed it already.  I just did the same change locally and have pushed it.

mo-nathan committed:
Adding SpeciesList as a language tag because of the GitHub CI/CD system not because I understand why it thinks we need it

mo-nathan committed:
Fix bad fixture

mo-nathan commented:
I don't understand the errors about :SpeciesList not being a language tag.  I don't see where it ever gets used, but I added it and we'll see if it resolves the issue.

JoeCohen commented:
`localhost:3000/qr/6F-0001` Last Observation is a surprise.
I expected it to show my last Observation.
Instead it's showing the Obs at the top of the Activity Feed, which:
- was created 2013-06-23, 
- updated 2024-03-20 06:34:14 PDT (-0700)
- created by Vlad Lekach (laddycans)
- for which I proposed a Name
- is the local copy of https://mushroomobserver.org/137413
<img width="339" alt="Screenshot 2024-03-23 at 12 46 18 PM" src="https://github.com/MushroomObserver/mushroom-observer/assets/45460/095acd30-e98a-4686-a144-389963f67435">

JoeCohen commented:
My "fix" just exposed more issues. Sorry. (I forgot to re-run the system tests.)

mo-nathan commented:
Looks like I didn't remember how to run the tests in test/system/.

JoeCohen commented on app/views/field_slips/new.html.erb:
```diff
@@ -0,0 +1,9 @@
+<h1>New field slip</h1>
```
1. Consider a different title; I think many (most?) people will confuse a physical slip with a FieldSlip object, and wonder why they're making a "new" slip when they have one in front of them. 
But I don't have a great suggestion. Maybe "Link Observation to Field Slip".
2. Whatever the title, use a "translation". (Even if you don't anticipate use in forays outside the USA, some foray participants -- particularly at NAMA -- may not be native English speakers.)

mo-nathan replied:
Interesting thought.  Thanks for catching the missing translation.  This naming came from the scaffold.  I generally need to fix up titles on the various pages.

nimmolo commented on app/helpers/field_slips_helper.rb:
```diff
@@ -0,0 +1,9 @@
+
+    ObservationView.last(User.current)
+  end
+end
```
This is potentially a generally useful helper, but just remember that all helpers are in a global "helper namespace". If it's just for field slips, you may want a more specific name.

mo-nathan replied:
I think this could be helpful in other places and it is not specific to FieldSlips.  I evolved it to create ObservationView.last(User) which might be a better standard to use than having it as a helper.  I don't have a strong opinion here.

nimmolo commented:
> `localhost:3000/qr/6F-0001` Last Observation is a surprise. I expected it to show my last Observation. Instead it's showing the Obs at the top of the Activity Feed, which:
> 
> * was created 2013-06-23,
> * updated 2024-03-20 06:34:14 PDT (-0700)
> * created by Vlad Lekach (laddycans)
> * for which I proposed a Name
> * is the local copy of https://mushroomobserver.org/137413

From checking through the new method in `ObservationView#last_observation`, I believe @mo-nathan has it pulling the _last viewed observation by the current user_, and that may not be the expected obs.

JoeCohen commented:
@nimmolo is right. field_slip#new is showing the last observation viewed by the current_user.

mo-nathan committed:
Bullet time

JoeCohen commented:
The `qr` url with Create Observation creates two Observations.
The first includes the field slip #, BUT omits images, the Project, Namings.
<img width="518" alt="Screenshot 2024-03-23 at 2 12 56 PM" src="https://github.com/MushroomObserver/mushroom-observer/assets/45460/2bff9636-bc9f-439b-b3f9-265cf3838bae">

The second is a more-or-less normal Observation, but omits the Field Slip.
<img width="1118" alt="Screenshot 2024-03-23 at 2 14 49 PM" src="https://github.com/MushroomObserver/mushroom-observer/assets/45460/d3159094-7316-4bc7-9532-368e1410d420">

JoeCohen commented:
PS: There's no log for the first Observation.

mo-nathan commented:
@JoeCohen Very interesting!  I know I manually tested after just giving a location and it seemed to work, but I didn't try adding images.  I'll look at it after I get back from a walk.

JoeCohen commented:
I just did the same thing, this time by scanning a QR code and not adding images.
Very slick and great to see it in action.
Unfortunately, same result: 2 Observations. 1st has FieldSlip but nothing else; 2nd has everything but the FieldSlip#.

JoeCohen commented:
@mo-nathan: I think it's a bug in constraint violations (or a bad interaction with that).
If I create an Observation that complies with the Project constraints, the above problem doesn't happen.
Each example I created above had a constraint violation, and I overrode the project constraints.

JoeCohen committed:
Test (red) exposing qr-new obs-violation bug

Exposes a bug when:
- scanning Field Slip
- choosing to Create New Observation
- completing Observation fields with Constraint Violation(s)
- then clicking Create

See https://github.com/MushroomObserver/mushroom-observer/pull/2024#issuecomment-2016604962 et seq.

JoeCohen committed:
Remove empty line at class body beginning

JoeCohen commented:
I pushed a short (failing) integration test exposing the bug. You can run it with:
`rails t test/integration/capybara/field_slips_integration_test.rb`
Sorry I didn't get any further.

If -- without using Field Slips -- I create an Observation, fill in the fields with something that violates project constraints, and click Create,
it: 
- redisplays the create_observation form with a flash warning ("This observation violates the constraints of one or more selected projects")
- does NOT create an Observation until I again click Create.

But if I scan a field slip, click Create New Observation, fill in constraint-violating info, then
- it redisplays the create_observation form, gives me the flash warning,
- BUT immediately creates an Observation (apparently including the FieldSlip, but ignoring namings, images).

mo-nathan commented:
To do:

- [ ] Fix double observation bug
- [ ] Fix titles (include a rename for the "New field slip" header)
- [ ] Fix system tests

mo-nathan commented:
> > `localhost:3000/qr/6F-0001` Last Observation is a surprise. I expected it to show my last Observation. Instead it's showing the Obs at the top of the Activity Feed, which:
> > 
> > * was created 2013-06-23,
> > * updated 2024-03-20 06:34:14 PDT (-0700)
> > * created by Vlad Lekach (laddycans)
> > * for which I proposed a Name
> > * is the local copy of https://mushroomobserver.org/137413
> 
> From checking through the new method in `ObservationView#last_observation`, I believe @mo-nathan has it pulling the _last viewed observation by the current user_, and that may not be the expected obs.

This is a really interesting case.  I can see an argument for limiting it to observations the current user created, but I think there is a legitimate use case where someone like a project admin might want to create a field slip object that the user creating the observation didn't know to create.  I'm hoping the use of the last observation feature is more of a rare power user thing and if they legitimately want to create a field slip for someone else's observation I think they should have a way to do that (e.g., a project admin sees an observation that should have been included that has a photo of the field slip, but the field slip link is on the observation).  Maybe I should add a comment on the page making it clearer that this is the last observation you viewed and if you view another observation in another window and reload the page or rescan the same QR code after visiting the observation you care about, then it will update.  Does that seem reasonable?

mo-nathan committed:
Start fixing test/system/field_slips_test.rb

mo-nathan committed:
Merge branch 'njw-field-slips' of github.com:MushroomObserver/mushroom-observer into njw-field-slips

mo-nathan committed:
Merge branch 'main' into njw-field-slips

mo-nathan committed:
Fix for double observation bug

mo-nathan commented:
> I pushed a short (failing) integration test exposing the bug. You can run it with: `rails t test/integration/capybara/field_slips_integration_test.rb` Sorry I didn't get any further.
> 
> If -- without using Field Slips -- I create an Observation, fill in the fields with something that violates project constraints, and click Create, it:
> 
> * redisplays the create_observation form with a flash warning ("This observation violates the constraints of one or more selected projects")
> * does NOT create an Observation until I again click Create.
> 
> But if I scan a field slip, click Create New Observation, fill in constraint-violating info, then
> 
> * it redisplays the create_observation form, gives me the flash warning,
> * BUT immediately creates an Observation (apparently including the FieldSlip, but ignoring namings, images).

Thanks for the test!  Turns out the fix was quite simple.  I was ignore the success flag.  I also need to pass the field_code param to the next page.

mo-nathan committed:
Fix field slip page titles and check that the observation has a thumbnail

mo-nathan committed:
Fix test/system/field_slips_test.rb tests

mo-nathan committed:
Rubocop

JoeCohen commented:
I banged on field slips for a while. 
A possible issue: inconsistent relation among FieldSlip, Observation, Project.
Here's what I did:
- Added a FS prefix to an existing project. (In my case I added `SARCO` to http://localhost:3000/projects/337.)
<img width="330" alt="Screenshot 2024-03-24 at 10 07 42 AM" src="https://github.com/MushroomObserver/mushroom-observer/assets/45460/f76950cd-0949-4f9c-a1ee-76a99ebc9e27">

- Created an Observation, **which was not in the Project**
- Pasted http://localhost:3000/qr/SARCO-0001 into the address bar, selected Last Observation (the one I just created). 

Result:
- Created a FieldSlip, correctly relating it to the Observation and the Project.
<img width="467" alt="Screenshot 2024-03-24 at 10 12 24 AM" src="https://github.com/MushroomObserver/mushroom-observer/assets/45460/70d6db43-991e-4110-a203-e82bfb86acbf">

- Updated the Observation, adding the FieldSlip #, but not the Project
<img width="322" alt="Screenshot 2024-03-24 at 10 18 17 AM" src="https://github.com/MushroomObserver/mushroom-observer/assets/45460/aa33d336-b55a-48b8-825f-290725809758">

- Did not add the Observation to the Project.

JoeCohen commented on app/controllers/observations_controller/new_and_create.rb:
```diff
@@ -102,6 +112,7 @@ def create
     success = false if @name && !@vote.value.nil? && !validate_object(@vote)
     success = false if @bad_images != []
     success = false if success && !save_observation(@observation)
+    update_field_slip(@observation, params[:field_code]) if success
```
Suggestions:
1. Put this line with the other things that get executed `if success`. (Result: one fewer conditional.)
2. Consider a guard clase.

The net result would be something like this:
```ruby
    success = false if success && !save_observation(@observation)
    return reload_new_form(params.dig(:naming, :reasons)) unless success

    # Once observation is saved we can save everything else.
    @observation.log(:log_observation_created)
    update_field_slip(@observation, params[:field_code])
    ...
    redirect_to_next_page
```

mo-nathan replied:
Nice!

nimmolo committed:
Adjust form templates, use simplified matrix boxes for "current"/"last"

nimmolo commented:
Attempted some improved markup for the form on #2063 - it's a PR off this branch

nimmolo committed:
Update _matrix_box.erb

Make header/footer more flexible

mo-nathan committed:
Merge branch 'main' into njw-field-slips

mo-nathan committed:
Tighten up new_and_create.rb code

nimmolo committed:
Use simpler markup, but DRY

nimmolo committed:
Update _form.html.erb

mo-nathan committed:
Merge branch 'njw-field-slips' into nimmo-field_slip-templates

mo-nathan committed:
Change the name of the link due to the page rearrangement

mo-nathan commented:
> I banged on field slips for a while. A possible issue: inconsistent relation among FieldSlip, Observation, Project. Here's what I did:
> 
> * Added a FS prefix to an existing project. (In my case I added `SARCO` to http://localhost:3000/projects/337.)
> 
> <img alt="Screenshot 2024-03-24 at 10 07 42 AM" width="330" src="https://private-user-images.githubusercontent.com/45460/316326658-f76950cd-0949-4f9c-a1ee-76a99ebc9e27.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTEzMDE4MDUsIm5iZiI6MTcxMTMwMTUwNSwicGF0aCI6Ii80NTQ2MC8zMTYzMjY2NTgtZjc2OTUwY2QtMDk0OS00ZjljLWExZWUtNzZhOTllYmM5ZTI3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMjQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzI0VDE3MzE0NVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTMxNDU0MTRkNjA0OGZkYTZiOTRkYjNlMDY2YjQxODM5ZmEwMjJhYTE5NGE1NjYyYjc5NjljODJhNGViZmE0NTMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.f8Or97bZ16F7DcFZmb8HNP4ou3bcTq4EeSHV90aPOzg">
> * Created an Observation, **which was not in the Project**
> * Pasted http://localhost:3000/qr/SARCO-0001 into the address bar, selected Last Observation (the one I just created).
> 
> Result:
> 
> * Created a FieldSlip, correctly relating it to the Observation and the Project:
> 
> <img alt="Screenshot 2024-03-24 at 10 12 24 AM" width="467" src="https://private-user-images.githubusercontent.com/45460/316327158-70d6db43-991e-4110-a203-e82bfb86acbf.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTEzMDE4MDUsIm5iZiI6MTcxMTMwMTUwNSwicGF0aCI6Ii80NTQ2MC8zMTYzMjcxNTgtNzBkNmRiNDMtOTkxZS00MTEwLWEyMDMtZTgyYmZiODZhY2JmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMjQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzI0VDE3MzE0NVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTRhZjc1ZTgzMGI3YjhiYWY2YzNjNGZlY2UwMDJmNDM0YjU3OGFiYWI2NmMzOWZlMTY0NDJkZjcxZGY3NDBlYjAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.n9Ujm6Hf67dp9V2gwyaguDEkhqTtl599oStB5PNHrCI">
> * Updated the Observation, adding the FieldSlip #, but not the Project:
> 
> <img alt="Screenshot 2024-03-24 at 10 18 17 AM" width="322" src="https://private-user-images.githubusercontent.com/45460/316327300-aa33d336-b55a-48b8-825f-290725809758.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTEzMDE4MDUsIm5iZiI6MTcxMTMwMTUwNSwicGF0aCI6Ii80NTQ2MC8zMTYzMjczMDAtYWEzM2QzMzYtYjU1YS00OGI4LTgyNWYtMjkwNzI1ODA5NzU4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMjQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzI0VDE3MzE0NVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTFhNjcwMmUwM2Q1ZTlhMTZiYTQyODEwMDI3MmY1MTY1YzJjMDU3NTBjOTIzZTU2ODc1MzU1NjY2MzNlMDIyNjYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.m4T0xSylOC91wQuwdil-8s27ABJ8rnlUFkZcFcayY_s">
> * Did not add the Observation to the Project.

Interesting case.  I see what you're driving at.  Howevver, I am imagining that observations associated with field slips that are not associated with the same project to be reasonable common in the long run.  The case I'm thinking of are folks who go to events that are using MO field slips and walk away with a bunch of them.  In this case it seems reasonable to associate the field slip with the project (due to the prefix) and this could be useful once we have forms associated with projects, but if the observation doesn't meet the constraints of the project, then it's legit for it not to be added.  I guess it would be fine to add it to the project if it meets the contraints, but not if it doesn't with appropriate flash messages.

nimmolo committed:
Different order for new/edit, fix image link bug

mo-nathan committed:
Add last observation to field slip project if appropriate

mo-nathan commented:
Resolved last observation getting added to project.

mo-nathan committed:
Merge branch 'njw-field-slips' into nimmo-field_slip-templates

mo-nathan committed:
Merge pull request #2063 from MushroomObserver/nimmo-field_slip-templates

Field Slips - Adjust form templates
